internal/ndp/ndp.go

```diff
@@ -0,0 +1,217 @@
+	ns, ok := msg.(*ndp.NeighborSolicitation)
+	if !ok {
+		return dropReasonMessageType
+	}
+
+	// Retrieve the sender's source link-layer address.
+	var nsLLAddr net.HardwareAddr
+	for _, o := range ns.Options {
+		// Ignore other options, including target link-layer address instead of source.
+		lla, ok := o.(*ndp.LinkLayerAddress)
+		if !ok {
+			continue
+		}
+		if lla.Direction != ndp.Source {
+			continue
+		}
+
+		nsLLAddr = lla.Addr
+		break
+	}
+	if nsLLAddr == nil {
+		return dropReasonNoSourceLL
+	}
+
+	// Ignore ndp requests which do not indicate the target IP that we should announce.
+	if !a.Announce(ns.TargetAddress) {
+		return dropReasonAnnounceIP
+	}
+
+	// We are not the leader, do not reply.
+	if !a.Leader() {
+		return dropReasonNotLeader
+	}
+
+	// pkt.TargetIP has been vetted to be "the one".
+	glog.Infof("Request: who-has %s?  tell %s (%s). reply: %s is-at %s", ns.TargetAddress, src, nsLLAddr, ns.TargetAddress, a.hardwareAddr)
+
+	// This request was solicited.
+	solicited := true
+	if err := a.advertise(src, ns.TargetAddress, solicited); err != nil {
+		glog.Warningf("Failed to write NDP neighbor advertisement for %s: %s", ns.TargetAddress, err)
+	}
+
+	return dropReasonNone
+}
+
+// advertise sends a NDP neighbor advertisement to dst for IP target using the
+// client in a.
+func (a *Announce) advertise(dst, target net.IP, solicited bool) error {
+	m := &ndp.NeighborAdvertisement{
+		Solicited:     solicited,
+		TargetAddress: target,
+		Options: []ndp.Option{
+			&ndp.LinkLayerAddress{
+				Direction: ndp.Target,
+				Addr:      a.hardwareAddr,
+			},
+		},
+	}
+
+	return a.conn.WriteTo(m, nil, dst)
+}
+
+// Close closes the ndp client in a.
+func (a *Announce) Close() error {
+	return a.conn.Close()
+}
+
+// SetBalancer adds ip to the set of announced addresses.
+func (a *Announce) SetBalancer(name string, ip net.IP) {
+	a.Lock()
+	defer a.Unlock()
+	a.ips[name] = ip
+}
+
+// DeleteBalancer deletes an address from the set of addresses we should announce.
+func (a *Announce) DeleteBalancer(name string) {
+	a.Lock()
+	defer a.Unlock()
+	if _, ok := a.ips[name]; ok {
+		delete(a.ips, name)
+	}
+}
+
+// Announce checks if ip should be announced.
+func (a *Announce) Announce(ip net.IP) bool {
+	a.RLock()
+	defer a.RUnlock()
+	for _, i := range a.ips {
+		if i.Equal(ip) {
+			return true
+		}
+	}
+	return false
+}
+
+// AnnounceName returns true when we have an announcement under name.
+func (a *Announce) AnnounceName(name string) bool {
+	a.RLock()
+	defer a.RUnlock()
+	_, ok := a.ips[name]
+	return ok
+}
+
+// Advertise sends unsolicited NDP neighbor advertisements for all IPs.
+func (a *Announce) Advertise() {
+	a.RLock()
+	defer a.RUnlock()
+
+	solicited := false
+	for _, ip := range a.ips {
+		_ = a.advertise(net.IPv6linklocalallnodes, ip, solicited)
+	}
+}
+
+// interfaceByIP returns the interface that has ip.
+func interfaceByIP(ip net.IP) (*net.Interface, error) {
```

Comment:
As a followup we should start with moving this to internal/interface so both arp and ndp can use it.

Reply:
Agreed.  I made no effort to dedup as I assume @danderson has put more thought into the abstractions.